plugins/obs-webrtc/webrtc/src/output.rs

```diff
@@ -0,0 +1,405 @@
+            ..Default::default()
+        }));
+
+        let (worker_tx, mut worker_rx): (UnboundedSender<Message>, UnboundedReceiver<Message>) =
+            mpsc::unbounded_channel();
+        let error_callback: Arc<Mutex<Option<ErrorCallback>>> = Arc::new(Mutex::new(None));
+
+        let mut interval = interval(Duration::from_millis(500));
+        let worker_future: JoinHandle<()> = std::thread::spawn({
+            let audio_track = audio_track.clone();
+            let video_track = video_track.clone();
+            let peer_connection = peer_connection.clone();
+            let stats = stats.clone();
+            let error_callback = error_callback.clone();
+            move || {
+                let runtime = tokio::runtime::Builder::new_multi_thread()
+                    .enable_all()
+                    .thread_name("webrtc_worker")
+                    .worker_threads(2)
+                    .build()
+                    .unwrap();
+                let runtime = Arc::new(runtime);
+                runtime.block_on({
+                    let runtime = runtime.clone();
+                    async move {
+                        'worker: loop {
+                            select! {
+                                message = worker_rx.recv() => {
+                                    match message {
+                                        Some(Message::Packet(packet)) => {
+                                            let sample = Sample {
+                                                data: Bytes::from(packet.data),
+                                                duration: packet.duration,
+                                                ..Default::default()
+                                            };
+                                            match (packet.typ, peer_connection.connection_state()) {
+                                                (_, s @ RTCPeerConnectionState::Failed | s @ RTCPeerConnectionState::Closed) => {
+                                                    error!("Invalid connection state for write_video: {s}");
+
+                                                    // Close worker thread so callback doesn't trigger a message to the worker thread
+                                                    worker_rx.close();
+                                                    runtime.spawn_blocking(move || {
+                                                        if let Some(callback) = &mut *error_callback.lock().unwrap() {
+                                                            callback(OutputStreamError::NetworkError);
+                                                        }
+                                                    }).await.unwrap_or_else(|e| {
+                                                        error!("Failed executing error callback: {e}");
+                                                    });
+
+                                                    break 'worker;
```

Comment:
I think it would make more sense to move this error handling, which is duplicated across most branches and deeply nested, below the loop, and make the loop `break` a boolean which indicates error exit vs non-error exit. Currently it is pretty hard to get an overview over this code segment because it is cluttered with repetitive error handling.

Reply:
Yeah, as I was rewriting this it ended up spinning out like that, I'll clean it up

Reply:
Fixed in d22a756afef63a9c504fd1bd32eb302810ca6fd2